### PR TITLE
SONARPY-2010 Avoid creating duplicate types when the corresponding symbols are duplicated

### PR DESCRIPTION
**Pull request by guillaume-dequenne-sonarsource**

### CODE REVIEW

File: python-checks/src/test/resources/checks/datetime_constructor_use_timezone_aware.py [hunk 1]
```diff
@@ -36,8 +36,5 @@ def aliased_utcnow():
     # FN because we lose the FQN in the assignment
     reassigned()
 
-    from datetime.datetime import utcnow as aliased
-    aliased() # Noncompliant
-
-    from datetime.datetime import utcnow
-    utcnow() # Noncompliant
+    from datetime import datetime as aliased
+    aliased.utcnow() # Noncompliant
```
Comment (guillaume-dequenne-sonarsource) [inline python-checks/src/test/resources/checks/datetime_constructor_use_timezone_aware.py:40]: This triggered an edge case when replacing members of `ModuleType` because we'd end up replacing the `datetime.datetime` `ClassType` with an empty `ModuleType` (after unsuccessfully trying to resolve it as a module).

File: python-frontend/src/main/java/org/sonar/python/semantic/v2/SymbolsModuleTypeProvider.java [hunk 2]
```diff
@@ -69,6 +75,52 @@ public ModuleType createModuleType(List<String> moduleFqn, ModuleType parent) {
...
+      }
+      // We have resolved an enclosing module, continue searching
+      currentModule = moduleType;
+    }
+    return PythonType.UNKNOWN;
+  }
+
+  public PythonType resolveLazyType(LazyType lazyType) {
+    if (lazyTypesContext.resolvedLazyTypes().containsKey(lazyType)) {
+      return lazyTypesContext.resolvedLazyTypes().get(lazyType);
```
Comment (guillaume-dequenne-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/semantic/v2/SymbolsModuleTypeProvider.java:104]: This line is uncovered, and I think because `FunctionType#returnType` now automatically resolves lazy types when called, I don't think it can be triggered anymore. Still, I prefer to keep this defensively, as it will be possible to trigger it when we resolve more types as lazy types.

File: python-frontend/src/main/java/org/sonar/python/types/v2/FunctionType.java [hunk 3]
```diff
@@ -98,6 +98,9 @@ public List<ParameterV2> parameters() {
   }
 
   public PythonType returnType() {
+    if (returnType instanceof LazyType lazyType) {
```
Comment (maksim-grebeniuk-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/types/v2/FunctionType.java:101]: wrap it to some utils, since we'll needed in a lot of places once we'll introduce lazy types for other places like superclasses

File: python-frontend/src/main/java/org/sonar/python/types/v2/FunctionType.java [hunk 4]
```diff
@@ -121,4 +124,11 @@ public boolean hasVariadicParameter() {
   public PythonType owner() {
     return owner;
   }
+
+  public void resolveLazyReturnType(PythonType pythonType) {
+    if (!(returnType instanceof LazyType)) {
```
Comment (maksim-grebeniuk-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/types/v2/FunctionType.java:129]: same here, possibly we may want to move this check to utils

File: python-frontend/src/main/java/org/sonar/python/types/v2/LazyType.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  public PythonType resolve() {
+    PythonType resolvedType = symbolsModuleTypeProvider.resolveLazyType(this);
+    consumers.forEach(c -> c.accept(resolvedType));
+    consumers.clear();
+    return resolvedType;
+  }
+
+  @Override
+  public PythonType unwrappedType() {
+    return this.resolve().unwrappedType();
```
Comment (maksim-grebeniuk-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/types/v2/LazyType.java:65]: since the current idea is that we never operate with lazy type outside of TypeInference, possibly we don't need to override this methods or need to throw an exception here

File: python-frontend/src/main/java/org/sonar/python/types/v2/ModuleType.java [hunk 6]
```diff
@@ -36,6 +36,14 @@ public ModuleType(@Nullable String name, @Nullable ModuleType parent, Map<String
     this.name = name;
     this.parent = parent;
     this.members = members;
+    if (this.parent == null) {
+      return;
```
Comment (maksim-grebeniuk-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/types/v2/ModuleType.java:40]: having an early return in a simple constructor confused me a bit from understanding what's going on here. Maybe you can extend this logic to a private method like `registerToParent` or revert the condition

File: python-frontend/src/main/java/org/sonar/python/types/v2/TypeCheckBuilder.java [hunk 7]
```diff
@@ -116,6 +116,9 @@ public IsSameAsTypePredicate(PythonType expectedType) {
 
     @Override
     public TriBool test(PythonType pythonType) {
+      if (pythonType instanceof LazyType lazyType) {
+        pythonType = lazyType.resolve();
```
Comment (maksim-grebeniuk-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/types/v2/TypeCheckBuilder.java:120]: not needed since lazy type should not be accessible outside of type inference

File: python-frontend/src/main/java/org/sonar/python/semantic/v2/LazyTypesContext.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Queue;
+import org.sonar.python.types.v2.LazyType;
+import org.sonar.python.types.v2.PythonType;
+
+public class LazyTypesContext {
+  private final Map<String, LazyType> lazyTypes;
+  private final Map<LazyType, Queue<String>> lazyTypeRequirements;
```
Comment (maksim-grebeniuk-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/semantic/v2/LazyTypesContext.java:30]: I guess this information could be inferred from LazyType itself

File: python-frontend/src/main/java/org/sonar/python/types/v2/LazyType.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+import java.util.Queue;
+import java.util.function.Consumer;
+import org.sonar.plugins.python.api.LocationInFile;
+import org.sonar.python.semantic.v2.SymbolsModuleTypeProvider;
+
+public class LazyType implements PythonType {
+
+  String fullyQualifiedName;
+  private final Queue<Consumer<PythonType>> consumers;
+  private final SymbolsModuleTypeProvider symbolsModuleTypeProvider;
```
Comment (maksim-grebeniuk-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/types/v2/LazyType.java:33]: I think it is better to operate with ProjectLevelTypesTable cause once you trigger lazy type resolution - it's type could be already resolved

File: python-frontend/src/main/java/org/sonar/python/semantic/v2/SymbolsModuleTypeProvider.java [hunk 10]
```diff
@@ -140,16 +195,63 @@ private PythonType resolveReturnType(Map<Symbol, PythonType> createdTypesBySymbo
...
-      .orElseGet(() -> convertToType(classSymbol, createdTypesBySymbol));
+    return resolvePossibleLazyType(createdTypesBySymbol, classSymbol, fullyQualifiedName);
+  }
+
+  PythonType resolvePossibleLazyType(Map<Symbol, PythonType> createdTypesBySymbol, ClassSymbol classSymbol, String fullyQualifiedName) {
+    PythonType currentModule = rootModule;
+    String[] split = fullyQualifiedName.split("\\.");
+    StringBuilder sb = new StringBuilder();
+    for (int i = 0; i < split.length; i++) {
+      String currentElement = split[i];
```
Comment (maksim-grebeniuk-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/semantic/v2/SymbolsModuleTypeProvider.java:213]: I think this logic is somehow implemented in `ProjectLevelTypeTable`, so maybe what we just need is to have a link to the types table in lazy type and call `getType` or `getModule` from for unresolved module and then call resolution consumers. WDYT?

File: python-frontend/src/main/java/org/sonar/python/semantic/v2/SymbolsModuleTypeProvider.java (hunk 2 above)
Comment (maksim-grebeniuk-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/semantic/v2/SymbolsModuleTypeProvider.java:104]: as discussed this is not needed anymore

File: python-frontend/src/main/java/org/sonar/python/semantic/v2/SymbolsModuleTypeProvider.java [hunk 11]
```diff
@@ -69,6 +75,52 @@ public ModuleType createModuleType(List<String> moduleFqn, ModuleType parent) {
...
+      currentModule = moduleType;
+    }
+    return PythonType.UNKNOWN;
+  }
+
+  public PythonType resolveLazyType(LazyType lazyType) {
+    if (lazyTypesContext.resolvedLazyTypes().containsKey(lazyType)) {
+      return lazyTypesContext.resolvedLazyTypes().get(lazyType);
+    }
+    Queue<String> modules = lazyTypesContext.lazyTypeRequirements().get(lazyType);
```
Comment (maksim-grebeniuk-sonarsource) [inline python-frontend/src/main/java/org/sonar/python/semantic/v2/SymbolsModuleTypeProvider.java:106]: same here, if LazyType will have a link to project level type table then this  entire method is not needed

**Review by maksim-grebeniuk-sonarsource — CHANGES_REQUESTED** [review]
Looks good! I think it requires some changes (as discussed and mentioned in the comments) to get rid of duplicated logic for the modules resolution and unused `LazyType`'s methods

File: python-frontend/src/main/java/org/sonar/python/types/v2/FunctionType.java (hunk 4 above)
Comment (guillaume-dequenne-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/types/v2/FunctionType.java:129]: I guess this case is a bit different since `resolveLazyReturnType` is essentially a setter that can be triggered only if the current return type is a lazy type. I still want to prevent using it if the current return type is not a lazy type.

File: python-frontend/src/main/java/org/sonar/python/types/v2/LazyType.java (hunk 9 above)
Comment (guillaume-dequenne-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/types/v2/LazyType.java:33]: I'm still not sure on that one, because `LazyType` themselves are created in `SymbolsModuleTypeProvider` during some actual conversions from symbols to types.

While the current architecture is already a bit odd, I feel it would be even weirder if `SymbolsModuleTypeProvider` held a reference to `ProjectLevelTypeTable` simply for the sake of passing it to `LazyType`. I don't really have a strong opinion here though; and.I feel this is temporary code anyway since it will only exist while we are doing symbol to type conversions.

File: python-frontend/src/main/java/org/sonar/python/semantic/v2/SymbolsModuleTypeProvider.java (hunk 10 above)
Comment (guillaume-dequenne-sonarsource) [reply python-frontend/src/main/java/org/sonar/python/semantic/v2/SymbolsModuleTypeProvider.java:210]: I moved the `getType` implementation in `SymbolsModuleTypeProvider` (for the reason highlighted in my other comment: that I'm not sure it's a good idea to pass `ProjectLevelTypeTable` to `SymbolsModuleTypeProvider` then to `LazyType`). We can discuss this further though.

**Review by maksim-grebeniuk-sonarsource — CHANGES_REQUESTED** [review]
From the current implementation perspective the code looks good.
As we discussed I guess the implementation will keep proper responsibilities if the implementation will look like:
## getModule lazy type workflow
1. `ProjectLevelTypeTable.getModule()`
2. Create module type with members
2.2. For a return type create a lazy type without any checks for existing types, add them to LazyContext
2.3 return module type
3. check if newly created lazy types already have types in the type table, resolve for existing ones
4. check if there are previously created lazy types exist for just created module type, resolve it
5. if there is no requested module  - then set a type of all lazy types assigned to this module to `UNKNOWN`
6. if there is a module but no member for lazy type - set its type to unknown

## LazyType resolving when accessed
1. call projectLevelTypeTable.getModule()
2. if a type is still not resolved, resolve it to UNKNOWN

As result: 
- SymbolModuleTypeProvider will be still just a convertor from Symbol to Type model
- Logic of LazyType creation will be simple without any checks and conditions
- LazyTypes resolution will operate with types API independent on symbols layer

**Review by maksim-grebeniuk-sonarsource — APPROVED** [review]
LGTM, great job, thanks!
As discussed, the only change needed is to rewrite and clean up a bit the `SymbolsModuleTypeProvider.resolvePossibleLazyType` method
